File: modules/songseed-pitch-shift/ios/SongseedPitchShiftModule.swift
import AVFoundation
import ExpoModulesCore

public class SongseedPitchShiftModule: Module {
  private lazy var engine = SongseedPitchShiftPracticeEngine(
    onStateChange: { payload in
      self.sendEvent("onStateChange", payload)
    },
    onPlaybackEnded: { payload in
      self.sendEvent("onPlaybackEnded", payload)
    },
    onError: { message in
      self.sendEvent("onError", ["message": message])
    }
  )
  private lazy var renderer = SongseedPitchShiftRenderer()

  public func definition() -> ModuleDefinition {
    Name("SongseedPitchShift")

    Events("onStateChange", "onPlaybackEnded", "onError")

    Function("isAvailable") {
      true
    }

    AsyncFunction("getCapabilities") { () -> [String: Any] in
      return self.engine.getCapabilities()
    }

    AsyncFunction("getPlaybackState") { () -> [String: Any] in
      return self.engine.getState()
    }

    AsyncFunction("loadForPractice") { (request: [String: Any]) -> [String: Any] in
      return try self.engine.loadForPractice(request)
    }

    AsyncFunction("play") { () -> [String: Any] in
      return try self.engine.play()
    }

    AsyncFunction("pause") { () -> [String: Any] in
      return self.engine.pause()
    }

    AsyncFunction("stop") { () -> [String: Any] in
      return self.engine.stop()
    }

    AsyncFunction("unload") { () -> [String: Any] in
      return self.engine.unload()
    }

    AsyncFunction("seekTo") { (positionMs: Double) -> [String: Any] in
      return try self.engine.seekTo(positionMs)
    }

    AsyncFunction("setPlaybackRate") { (rate: Double) -> [String: Any] in
      return self.engine.setPlaybackRate(rate)
    }

    AsyncFunction("setPitchShiftSemitones") { (semitones: Int) -> [String: Any] in
      return self.engine.setPitchShiftSemitones(semitones)
    }

    AsyncFunction("renderPitchShiftedFile") { (request: [String: Any]) -> [String: Any] in
      return try self.renderer.renderFile(request)
    }

    AsyncFunction("renderMixedFile") { (request: [String: Any]) -> [String: Any] in
      return try self.renderer.renderMixedFile(request)
    }

    OnDestroy {
      _ = self.engine.unload()
    }
  }
}

File: modules/songseed-pitch-shift/android/src/main/java/expo/modules/songseedpitchshift/SongseedPitchShiftModule.kt
package expo.modules.songseedpitchshift

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class SongseedPitchShiftModule : Module() {
  private val engine by lazy {
    SongseedPitchShiftPracticeEngine(
      context = appContext.reactContext ?: throw IllegalStateException("React context unavailable"),
      onStateChange = { payload ->
        sendEvent("onStateChange", payload)
      },
      onPlaybackEnded = { payload ->
        sendEvent("onPlaybackEnded", payload)
      },
      onError = { message ->
        sendEvent("onError", mapOf("message" to message))
      },
    )
  }
  private val renderer by lazy {
    SongseedPitchShiftRenderer(
      context = appContext.reactContext ?: throw IllegalStateException("React context unavailable"),
    )
  }

  override fun definition() = ModuleDefinition {
    Name("SongseedPitchShift")

    Events("onStateChange", "onPlaybackEnded", "onError")

    Function("isAvailable") {
      true
    }

    AsyncFunction("getCapabilities") {
      engine.getCapabilities()
    }

    AsyncFunction("getPlaybackState") {
      engine.getState()
    }

    AsyncFunction("loadForPractice") { request: Map<String, Any?> ->
      engine.loadForPractice(request)
    }

    AsyncFunction("play") {
      engine.play()
    }

    AsyncFunction("pause") {
      engine.pause()
    }

    AsyncFunction("stop") {
      engine.stop()
    }

    AsyncFunction("unload") {
      engine.unload()
    }

    AsyncFunction("seekTo") { positionMs: Double ->
      engine.seekTo(positionMs)
    }

    AsyncFunction("setPlaybackRate") { rate: Double ->
      engine.setPlaybackRate(rate)
    }

    AsyncFunction("setPitchShiftSemitones") { semitones: Int ->
      engine.setPitchShiftSemitones(semitones)
    }

    AsyncFunction("renderPitchShiftedFile") { request: Map<String, Any?> ->
      renderer.renderFile(request)
    }

    AsyncFunction("renderMixedFile") { request: Map<String, Any?> ->
      renderer.renderMixedFile(request)
    }

    AsyncFunction("renderTrim") { request: Map<String, Any?> ->
      renderer.renderTrim(request)
    }

    AsyncFunction("computeWaveform") { request: Map<String, Any?> ->
      renderer.computeWaveform(request)
    }

    // Preempt in-flight/queued waveform decodes carrying an epoch older than `epoch`.
    // Called by JS when playback starts so the decoder frees the MediaCodec pool.
    Function("cancelActiveWaveform") { epoch: Double ->
      renderer.cancelActiveWaveform(epoch.toLong())
    }

    OnDestroy {
      engine.unload()
    }
  }
}
